File: memcontainers/shcore/src/arith.rs
//! POSIX arithmetic expansion `$(( … ))`.
//!
//! A pure integer (i64) evaluator with C-like precedence (precedence-climbing).
//! OS-free: variable reads and writes go through an [`ArithEnv`] the caller supplies,
//! so the evaluator is trivially unit-testable and the executor wires it to the shell
//! variable map. Integer semantics are wrapping (two's-complement), matching the shell.
//!
//! The single-trait env (rather than two `FnMut` closures over the same map) is what
//! lets the executor implement `get`/`set` with one ordinary borrow — no raw pointers,
//! no `unsafe`.

use alloc::string::{String, ToString};
use alloc::vec::Vec;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ArithError(pub String);

/// Variable access for the arithmetic evaluator. `get` returns 0 for an unset or
/// non-numeric name; `set` writes back (for `=`, `+=`, …, `++`, `--`).
pub trait ArithEnv {
    fn get(&mut self, name: &str) -> i64;
    fn set(&mut self, name: &str, val: i64);
}

/// Evaluate `expr` against `env`.
pub fn eval(expr: &str, env: &mut dyn ArithEnv) -> Result<i64, ArithError> {
    let toks = lex(expr)?;
    let mut p = Parser { toks, pos: 0, env };
    let v = p.expr(0)?;
    p.expect_eof()?;
    Ok(v)
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum Tok {
    Num(i64),
    Name(String),
    Op(&'static str),
    LParen,
    RParen,
}

fn lex(s: &str) -> Result<Vec<Tok>, ArithError> {
    let c: Vec<char> = s.chars().collect();
    let mut i = 0;
    let mut out = Vec::new();
    while i < c.len() {
        let ch = c[i];
        if ch.is_ascii_whitespace() {
            i += 1;
            continue;
        }
        if ch.is_ascii_digit() {
            // decimal, 0x.. hex, 0.. octal
            let start = i;
            if ch == '0' && i + 1 < c.len() && (c[i + 1] == 'x' || c[i + 1] == 'X') {
                i += 2;
                while i < c.len() && c[i].is_ascii_hexdigit() {
                    i += 1;
                }
                let hex: String = c[start + 2..i].iter().collect();
                let v = i64::from_str_radix(&hex, 16).map_err(|_| ArithError("bad hex".into()))?;
                out.push(Tok::Num(v));
                continue;
            }
            while i < c.len() && c[i].is_ascii_digit() {
                i += 1;
            }
            let dec: String = c[start..i].iter().collect();
            let v: i64 = dec.parse().map_err(|_| ArithError("bad number".into()))?;
            out.push(Tok::Num(v));
            continue;
        }
        if ch == '_' || ch.is_ascii_alphabetic() {
            let start = i;
            while i < c.len() && (c[i] == '_' || c[i].is_ascii_alphanumeric()) {
                i += 1;
            }
            out.push(Tok::Name(c[start..i].iter().collect()));
            continue;
        }
        if ch == '(' {
            out.push(Tok::LParen);
            i += 1;
            continue;
        }
        if ch == ')' {
            out.push(Tok::RParen);
            i += 1;
            continue;
        }
        // multi-char operators (maximal munch)
        let two: String = c[i..(i + 2).min(c.len())].iter().collect();
        let three: String = c[i..(i + 3).min(c.len())].iter().collect();
        const OPS3: &[&str] = &["<<=", ">>="];
        const OPS2: &[&str] = &[
            "<<", ">>", "<=", ">=", "==", "!=", "&&", "||", "++", "--", "+=", "-=", "*=", "/=",
            "%=", "&=", "|=", "^=",
        ];
        if let Some(op) = OPS3.iter().find(|o| **o == three) {
            out.push(Tok::Op(op));
            i += 3;
            continue;
        }
        if let Some(op) = OPS2.iter().find(|o| **o == two) {
            out.push(Tok::Op(op));
            i += 2;
            continue;
        }
        const OPS1: &[&str] = &[
            "+", "-", "*", "/", "%", "<", ">", "=", "!", "~", "&", "|", "^", "?", ":", ",",
        ];
        let one: String = ch.to_string();
        if let Some(op) = OPS1.iter().find(|o| **o == one) {
            out.push(Tok::Op(op));
            i += 1;
            continue;
        }
        return Err(ArithError(alloc::format!("unexpected char '{ch}'")));
    }
    Ok(out)
}

struct Parser<'a> {
    toks: Vec<Tok>,
    pos: usize,
    env: &'a mut dyn ArithEnv,
}

impl Parser<'_> {
    fn peek(&self) -> Option<&Tok> {
        self.toks.get(self.pos)
    }
    fn bump(&mut self) -> Option<Tok> {
        let t = self.toks.get(self.pos).cloned();
        if t.is_some() {
            self.pos += 1;
        }
        t
    }
    fn expect_eof(&self) -> Result<(), ArithError> {
        if self.pos == self.toks.len() {
            Ok(())
        } else {
            Err(ArithError("trailing tokens in expression".into()))
        }
    }

    /// Precedence-climbing. `min_bp` is the minimum binding power to consume.
    fn expr(&mut self, min_bp: u8) -> Result<i64, ArithError> {
        let mut lhs = self.unary()?;
        loop {
            let op = match self.peek() {
                Some(Tok::Op(o)) => *o,
                _ => break,
            };
            // assignment (right-assoc, lowest) handled separately
            if let Some((lbp, rbp)) = infix_bp(op) {
                if lbp < min_bp {
                    break;
                }
                self.bump();
                if op == "?" {
                    // ternary: cond ? a : b
                    let then_v = self.expr(0)?;
                    match self.bump() {
                        Some(Tok::Op(":")) => {}
                        _ => return Err(ArithError("expected ':' in ?:".into())),
                    }
                    let else_v = self.expr(rbp)?;
                    lhs = if lhs != 0 { then_v } else { else_v };
                    continue;
                }
                let rhs = self.expr(rbp)?;
                lhs = apply(op, lhs, rhs)?;
                continue;
            }
            break;
        }
        Ok(lhs)
    }

    fn unary(&mut self) -> Result<i64, ArithError> {
        match self.peek().cloned() {
            Some(Tok::Op("+")) => {
                self.bump();
                self.unary()
            }
            Some(Tok::Op("-")) => {
                self.bump();
                Ok(-self.unary()?)
            }
            Some(Tok::Op("!")) => {
                self.bump();
                Ok((self.unary()? == 0) as i64)
            }
            Some(Tok::Op("~")) => {
                self.bump();
                Ok(!self.unary()?)
            }
            Some(Tok::Op("++")) | Some(Tok::Op("--")) => {
                let op = if let Some(Tok::Op(o)) = self.bump() {
                    o
                } else {
                    unreachable!()
                };
                // pre-inc/dec on a name
                if let Some(Tok::Name(n)) = self.peek().cloned() {
                    self.bump();
                    let cur = self.env.get(&n);
                    let nv = if op == "++" { cur + 1 } else { cur - 1 };
                    self.env.set(&n, nv);
                    Ok(nv)
                } else {
                    Err(ArithError("++/-- needs a variable".into()))
                }
            }
            _ => self.primary(),
        }
    }

    fn primary(&mut self) -> Result<i64, ArithError> {
        match self.bump() {
            Some(Tok::Num(n)) => Ok(n),
            Some(Tok::LParen) => {
                let v = self.expr(0)?;
                match self.bump() {
                    Some(Tok::RParen) => Ok(v),
                    _ => Err(ArithError("expected ')'".into())),
                }
            }
            Some(Tok::Name(n)) => {
                // assignment / compound-assignment / post inc-dec
                match self.peek().cloned() {
                    Some(Tok::Op("=")) => {
                        self.bump();
                        let v = self.expr(2)?; // assignment binds loosely, right-assoc
                        self.env.set(&n, v);
                        Ok(v)
                    }
                    Some(Tok::Op(op))
                        if matches!(
                            op,
                            "+=" | "-=" | "*=" | "/=" | "%=" | "<<=" | ">>=" | "&=" | "|=" | "^="
                        ) =>
                    {
                        self.bump();
                        let rhs = self.expr(2)?;
                        let cur = self.env.get(&n);
                        let base = &op[..op.len() - 1];
                        let v = apply(base, cur, rhs)?;
                        self.env.set(&n, v);
                        Ok(v)
                    }
                    Some(Tok::Op("++")) => {
                        self.bump();
                        let cur = self.env.get(&n);
                        self.env.set(&n, cur + 1);
                        Ok(cur)
                    }
                    Some(Tok::Op("--")) => {
                        self.bump();
                        let cur = self.env.get(&n);
                        self.env.set(&n, cur - 1);
                        Ok(cur)
                    }
                    _ => Ok(self.env.get(&n)),
                }
            }
            other => Err(ArithError(alloc::format!("unexpected token {other:?}"))),
        }
    }
}

/// Left/right binding powers for infix operators. Right-assoc ops use lbp==rbp+1.
fn infix_bp(op: &str) -> Option<(u8, u8)> {
    Some(match op {
        "," => (1, 2),
        "?" => (4, 3), // ternary, right-assoc
        "||" => (5, 6),
        "&&" => (7, 8),
        "|" => (9, 10),
        "^" => (11, 12),
        "&" => (13, 14),
        "==" | "!=" => (15, 16),
        "<" | "<=" | ">" | ">=" => (17, 18),
        "<<" | ">>" => (19, 20),
        "+" | "-" => (21, 22),
        "*" | "/" | "%" => (23, 24),
        _ => return None,
    })
}

fn apply(op: &str, a: i64, b: i64) -> Result<i64, ArithError> {
    Ok(match op {
        "," => b,
        "||" => ((a != 0) || (b != 0)) as i64,
        "&&" => ((a != 0) && (b != 0)) as i64,
        "|" => a | b,
        "^" => a ^ b,
        "&" => a & b,
        "==" => (a == b) as i64,
        "!=" => (a != b) as i64,
        "<" => (a < b) as i64,
        "<=" => (a <= b) as i64,
        ">" => (a > b) as i64,
        ">=" => (a >= b) as i64,
        "<<" => a.wrapping_shl(b as u32),
        ">>" => a.wrapping_shr(b as u32),
        "+" => a.wrapping_add(b),
        "-" => a.wrapping_sub(b),
        "*" => a.wrapping_mul(b),
        "/" => {
            if b == 0 {
                return Err(ArithError("division by zero".into()));
            }
            a.wrapping_div(b)
        }
        "%" => {
            if b == 0 {
                return Err(ArithError("division by zero".into()));
            }
            a.wrapping_rem(b)
        }
        _ => return Err(ArithError(alloc::format!("bad operator '{op}'"))),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::collections::BTreeMap;

    /// A simple in-memory env for tests.
    struct Vars(BTreeMap<String, i64>);
    impl ArithEnv for Vars {
        fn get(&mut self, name: &str) -> i64 {
            *self.0.get(name).unwrap_or(&0)
        }
        fn set(&mut self, name: &str, val: i64) {
            self.0.insert(name.to_string(), val);
        }
    }

    fn ev(s: &str) -> i64 {
        let mut env = Vars(BTreeMap::new());
        eval(s, &mut env).unwrap()
    }

    #[test]
    fn precedence_and_parens() {
        assert_eq!(ev("1 + 2 * 3"), 7);
        assert_eq!(ev("(1 + 2) * 3"), 9);
        assert_eq!(ev("2 * 3 % 4"), 2);
        assert_eq!(ev("10 - 2 - 3"), 5); // left-assoc
    }

    #[test]
    fn comparisons_logical_ternary() {
        assert_eq!(ev("1 < 2"), 1);
        assert_eq!(ev("2 <= 1"), 0);
        assert_eq!(ev("1 && 0"), 0);
        assert_eq!(ev("1 || 0"), 1);
        assert_eq!(ev("(5 > 3) ? 10 : 20"), 10);
        assert_eq!(ev("0 ? 10 : 20"), 20);
    }

    #[test]
    fn bitwise_shift_unary_hex() {
        assert_eq!(ev("1 << 4"), 16);
        assert_eq!(ev("0xff & 0x0f"), 15);
        assert_eq!(ev("~0"), -1);
        assert_eq!(ev("-5 + 3"), -2);
        assert_eq!(ev("!0"), 1);
    }

    #[test]
    fn variables_and_assignment() {
        let mut env = Vars(BTreeMap::new());
        env.0.insert("x".into(), 5);
        assert_eq!(eval("x * 2 + 1", &mut env).unwrap(), 11);
    }

    #[test]
    fn assignment_writes_back_through_env() {
        let mut env = Vars(BTreeMap::new());
        // One env handles both reads and writes — no aliasing gymnastics.
        assert_eq!(eval("y = 3 + 4", &mut env).unwrap(), 7);
        assert_eq!(env.0.get("y"), Some(&7));
        // Compound assignment and pre/post inc read-modify-write the same env.
        assert_eq!(eval("y += 3", &mut env).unwrap(), 10);
        assert_eq!(eval("++y", &mut env).unwrap(), 11);
        assert_eq!(eval("y++", &mut env).unwrap(), 11);
        assert_eq!(env.0.get("y"), Some(&12));
    }

    #[test]
    fn div_by_zero_errors() {
        let mut env = Vars(BTreeMap::new());
        assert!(eval("1 / 0", &mut env).is_err());
    }
}

File: memcontainers/shcore/src/arith.zig
//! POSIX arithmetic expansion `$(( ... ))`.

const std = @import("std");

pub const ArithError = error{
    BadHex,
    BadNumber,
    BadOperator,
    DivisionByZero,
    ExpectedColon,
    ExpectedRParen,
    NeedsVariable,
    OutOfMemory,
    TrailingTokens,
    UnexpectedChar,
    UnexpectedToken,
};

pub const ArithEnv = struct {
    ptr: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        get: *const fn (*anyopaque, []const u8) i64,
        set: *const fn (*anyopaque, []const u8, i64) void,
    };

    pub fn get(self: *ArithEnv, name: []const u8) i64 {
        return self.vtable.get(self.ptr, name);
    }

    pub fn set(self: *ArithEnv, name: []const u8, val: i64) void {
        self.vtable.set(self.ptr, name, val);
    }
};

pub fn eval(allocator: std.mem.Allocator, expr: []const u8, env: *ArithEnv) ArithError!i64 {
    const toks = try lex(allocator, expr);
    var parser = Parser{ .toks = toks, .pos = 0, .env = env };
    const value = try parser.expr(0);
    try parser.expectEof();
    return value;
}

const Tok = union(enum) {
    num: i64,
    name: []const u8,
    op: []const u8,
    lparen,
    rparen,
};

fn lex(allocator: std.mem.Allocator, source: []const u8) ArithError![]Tok {
    var out = std.ArrayList(Tok).empty;
    var i: usize = 0;
    while (i < source.len) {
        const ch = source[i];
        if (std.ascii.isWhitespace(ch)) {
            i += 1;
            continue;
        }
        if (std.ascii.isDigit(ch)) {
            const start = i;
            if (ch == '0' and i + 1 < source.len and (source[i + 1] == 'x' or source[i + 1] == 'X')) {
                i += 2;
                const hex_start = i;
                while (i < source.len and std.ascii.isHex(source[i])) i += 1;
                const value = std.fmt.parseInt(i64, source[hex_start..i], 16) catch return error.BadHex;
                try out.append(allocator, .{ .num = value });
                continue;
            }
            while (i < source.len and std.ascii.isDigit(source[i])) i += 1;
            const value = std.fmt.parseInt(i64, source[start..i], 10) catch return error.BadNumber;
            try out.append(allocator, .{ .num = value });
            continue;
        }
        if (ch == '_' or std.ascii.isAlphabetic(ch)) {
            const start = i;
            i += 1;
            while (i < source.len and (source[i] == '_' or std.ascii.isAlphanumeric(source[i]))) i += 1;
            try out.append(allocator, .{ .name = source[start..i] });
            continue;
        }
        if (ch == '(') {
            try out.append(allocator, .lparen);
            i += 1;
            continue;
        }
        if (ch == ')') {
            try out.append(allocator, .rparen);
            i += 1;
            continue;
        }
        if (matchOp(source[i..])) |op| {
            try out.append(allocator, .{ .op = op });
            i += op.len;
            continue;
        }
        return error.UnexpectedChar;
    }
    return out.toOwnedSlice(allocator) catch return error.BadNumber;
}

fn matchOp(rest: []const u8) ?[]const u8 {
    inline for (.{ "<<=", ">>=" }) |op| {
        if (std.mem.startsWith(u8, rest, op)) return op;
    }
    inline for (.{ "<<", ">>", "<=", ">=", "==", "!=", "&&", "||", "++", "--", "+=", "-=", "*=", "/=", "%=", "&=", "|=", "^=" }) |op| {
        if (std.mem.startsWith(u8, rest, op)) return op;
    }
    inline for (.{ "+", "-", "*", "/", "%", "<", ">", "=", "!", "~", "&", "|", "^", "?", ":", "," }) |op| {
        if (std.mem.startsWith(u8, rest, op)) return op;
    }
    return null;
}

const Parser = struct {
    toks: []Tok,
    pos: usize,
    env: *ArithEnv,

    fn peek(self: *Parser) ?Tok {
        if (self.pos >= self.toks.len) return null;
        return self.toks[self.pos];
    }

    fn bump(self: *Parser) ?Tok {
        const tok = self.peek() orelse return null;
        self.pos += 1;
        return tok;
    }

    fn expectEof(self: *Parser) ArithError!void {
        if (self.pos != self.toks.len) return error.TrailingTokens;
    }

    fn expr(self: *Parser, min_bp: u8) ArithError!i64 {
        var lhs = try self.unary();
        while (true) {
            const op = switch (self.peek() orelse break) {
                .op => |o| o,
                else => break,
            };
            const bp = infixBp(op) orelse break;
            if (bp.l < min_bp) break;
            _ = self.bump();
            if (std.mem.eql(u8, op, "?")) {
                const then_v = try self.expr(0);
                switch (self.bump() orelse return error.ExpectedColon) {
                    .op => |colon| if (!std.mem.eql(u8, colon, ":")) return error.ExpectedColon,
                    else => return error.ExpectedColon,
                }
                const else_v = try self.expr(bp.r);
                lhs = if (lhs != 0) then_v else else_v;
                continue;
            }
            const rhs = try self.expr(bp.r);
            lhs = try apply(op, lhs, rhs);
        }
        return lhs;
    }

    fn unary(self: *Parser) ArithError!i64 {
        const tok = self.peek() orelse return error.UnexpectedToken;
        switch (tok) {
            .op => |op| {
                if (std.mem.eql(u8, op, "+")) {
                    _ = self.bump();
                    return self.unary();
                }
                if (std.mem.eql(u8, op, "-")) {
                    _ = self.bump();
                    return 0 -% try self.unary();
                }
                if (std.mem.eql(u8, op, "!")) {
                    _ = self.bump();
                    return if ((try self.unary()) == 0) 1 else 0;
                }
                if (std.mem.eql(u8, op, "~")) {
                    _ = self.bump();
                    return ~(try self.unary());
                }
                if (std.mem.eql(u8, op, "++") or std.mem.eql(u8, op, "--")) {
                    _ = self.bump();
                    const next = self.bump() orelse return error.NeedsVariable;
                    const name = switch (next) {
                        .name => |n| n,
                        else => return error.NeedsVariable,
                    };
                    const cur = self.env.get(name);
                    const nv = if (std.mem.eql(u8, op, "++")) cur +% 1 else cur -% 1;
                    self.env.set(name, nv);
                    return nv;
                }
            },
            else => {},
        }
        return self.primary();
    }

    fn primary(self: *Parser) ArithError!i64 {
        const tok = self.bump() orelse return error.UnexpectedToken;
        return switch (tok) {
            .num => |n| n,
            .lparen => blk: {
                const value = try self.expr(0);
                switch (self.bump() orelse return error.ExpectedRParen) {
                    .rparen => break :blk value,
                    else => return error.ExpectedRParen,
                }
            },
            .name => |name| try self.nameValue(name),
            else => error.UnexpectedToken,
        };
    }

    fn nameValue(self: *Parser, name: []const u8) ArithError!i64 {
        const tok = self.peek() orelse return self.env.get(name);
        const op = switch (tok) {
            .op => |o| o,
            else => return self.env.get(name),
        };
        if (std.mem.eql(u8, op, "=")) {
            _ = self.bump();
            const value = try self.expr(2);
            self.env.set(name, value);
            return value;
        }
        if (compoundBase(op)) |base| {
            _ = self.bump();
            const rhs = try self.expr(2);
            const value = try apply(base, self.env.get(name), rhs);
            self.env.set(name, value);
            return value;
        }
        if (std.mem.eql(u8, op, "++")) {
            _ = self.bump();
            const cur = self.env.get(name);
            self.env.set(name, cur +% 1);
            return cur;
        }
        if (std.mem.eql(u8, op, "--")) {
            _ = self.bump();
            const cur = self.env.get(name);
            self.env.set(name, cur -% 1);
            return cur;
        }
        return self.env.get(name);
    }
};

fn compoundBase(op: []const u8) ?[]const u8 {
    inline for (.{ "+=", "-=", "*=", "/=", "%=", "<<=", ">>=", "&=", "|=", "^=" }) |candidate| {
        if (std.mem.eql(u8, op, candidate)) return candidate[0 .. candidate.len - 1];
    }
    return null;
}

const BindingPower = struct { l: u8, r: u8 };

fn infixBp(op: []const u8) ?BindingPower {
    if (std.mem.eql(u8, op, ",")) return .{ .l = 1, .r = 2 };
    if (std.mem.eql(u8, op, "?")) return .{ .l = 4, .r = 3 };
    if (std.mem.eql(u8, op, "||")) return .{ .l = 5, .r = 6 };
    if (std.mem.eql(u8, op, "&&")) return .{ .l = 7, .r = 8 };
    if (std.mem.eql(u8, op, "|")) return .{ .l = 9, .r = 10 };
    if (std.mem.eql(u8, op, "^")) return .{ .l = 11, .r = 12 };
    if (std.mem.eql(u8, op, "&")) return .{ .l = 13, .r = 14 };
    if (std.mem.eql(u8, op, "==") or std.mem.eql(u8, op, "!=")) return .{ .l = 15, .r = 16 };
    if (std.mem.eql(u8, op, "<") or std.mem.eql(u8, op, "<=") or std.mem.eql(u8, op, ">") or std.mem.eql(u8, op, ">=")) return .{ .l = 17, .r = 18 };
    if (std.mem.eql(u8, op, "<<") or std.mem.eql(u8, op, ">>")) return .{ .l = 19, .r = 20 };
    if (std.mem.eql(u8, op, "+") or std.mem.eql(u8, op, "-")) return .{ .l = 21, .r = 22 };
    if (std.mem.eql(u8, op, "*") or std.mem.eql(u8, op, "/") or std.mem.eql(u8, op, "%")) return .{ .l = 23, .r = 24 };
    return null;
}

fn apply(op: []const u8, a: i64, b: i64) ArithError!i64 {
    if (std.mem.eql(u8, op, ",")) return b;
    if (std.mem.eql(u8, op, "||")) return if (a != 0 or b != 0) 1 else 0;
    if (std.mem.eql(u8, op, "&&")) return if (a != 0 and b != 0) 1 else 0;
    if (std.mem.eql(u8, op, "|")) return a | b;
    if (std.mem.eql(u8, op, "^")) return a ^ b;
    if (std.mem.eql(u8, op, "&")) return a & b;
    if (std.mem.eql(u8, op, "==")) return if (a == b) 1 else 0;
    if (std.mem.eql(u8, op, "!=")) return if (a != b) 1 else 0;
    if (std.mem.eql(u8, op, "<")) return if (a < b) 1 else 0;
    if (std.mem.eql(u8, op, "<=")) return if (a <= b) 1 else 0;
    if (std.mem.eql(u8, op, ">")) return if (a > b) 1 else 0;
    if (std.mem.eql(u8, op, ">=")) return if (a >= b) 1 else 0;
    if (std.mem.eql(u8, op, "<<")) return a << shiftAmount(b);
    if (std.mem.eql(u8, op, ">>")) return a >> shiftAmount(b);
    if (std.mem.eql(u8, op, "+")) return a +% b;
    if (std.mem.eql(u8, op, "-")) return a -% b;
    if (std.mem.eql(u8, op, "*")) return a *% b;
    if (std.mem.eql(u8, op, "/")) {
        if (b == 0) return error.DivisionByZero;
        if (a == std.math.minInt(i64) and b == -1) return std.math.minInt(i64);
        return @divTrunc(a, b);
    }
    if (std.mem.eql(u8, op, "%")) {
        if (b == 0) return error.DivisionByZero;
        if (a == std.math.minInt(i64) and b == -1) return 0;
        return @rem(a, b);
    }
    return error.BadOperator;
}

fn shiftAmount(v: i64) u6 {
    const as_u: u64 = @bitCast(v);
    return @truncate(as_u);
}
